README.md
# Secure Router

[![Build
Status](https://travis-ci.org/goodeggs/secure-router.svg?branch=master)](https://travis-ci.org/goodeggs/secure-router)

This node module implements a replacement for the `Router` that comes with
[Express][express]. It provides a stricter security model, making your
endpoints default to returning [`401 Unauthorized`][rfc401] or [`403
Forbidden`][rfc403] unless they have explicitly been **authenticated** and
**authorized**.

This module helps you avoid situations where you set up a route but forget
to add security middleware to it, leaving it unintentionally
world-readable/writeable. Using it is easy:

```javascript
import express from 'express';
import SecureRouter from 'secure-router';

const app = express();
const router = new SecureRouter();
router.bounceRequests();
app.use(router);

router.secureEndpoint({
  method: 'GET',
  path: '/the-crown-jewels',
  bouncers: [
    ((req) => if (req.user) return SecureRouter.AUTHENTICATE),
    ((req) => if (req.user.isRoyalty) return SecureRouter.AUTHORIZE),
  ],
  middleware (req, res) {
    req.send(theCrownJewels)
  },
});

/* OR */

secretRouter = router.secureSubpath({
  path: '/secrets',
  bouncers: [
    ((req) => if (req.user) return SecureRouter.AUTHENTICATE),
    ((req) => if (req.user.classification === 'MI6') return SecureRouter.AUTHORIZE),
    ((req) => if (req.user.number === '006') return return SecureRouter.DENY),
  ],
});

secretRouter.get('/mission-briefs', (req, res) => res.send(nextMission));
```

[rfc401]: https://httpstatuses.com/401
[rfc403]: https://httpstatuses.com/403
[express]: https://expressjs.com/

## Bouncers

You can define an arbitrary number of `bouncers` on routes, paths, or entire
routers.

Any bouncer can explicitly `DENY`, `AUTHENTICATE`, or `AUTHORIZE` a request for
a path for which that bouncer has been defined. This can be done asynchronously,
using a promise.

If any bouncer has elected to `DENY`, the request immediately short-circuits
with a `401` status.

If no bouncer has elected to `AUTHENTICATE`, the request immediately
short-circuits with a `401` status.

If no bouncer has elected to `AUTHORIZE`, the request immediately
short-circuits with a `403` status.

### Bouncer functions

All bouncers are individual functions, that accept a `req` and a `res` argument
(this is the same `req` that an express middleware receives), and return a
promise.  If the promise resolves to anything other than `SecureRouter.DENY`,
`SecureRouter.AUTHENTICATE`, `SecureRouter.AUTHORIZE`, or the result of
`SecureRouter.denyWith()`, the bouncer is ignored.

Example:

```js
const bouncer = function (req, res) {
  getSecurityClearanceLevel(req.user.id)
  .then(function (level) {
    if (level === 'TOP SECRET') return SecureRouter.AUTHORIZE;
    else return SecureRouter.denyWith({
      statusCode: 404,
      payload: 'Nothing to see here, move along.'
    });
  });
}
```

### `denyWith(<middleware|response>)`

Returns a bouncer response that will cause the bouncer to deny the request,
and will edit the response to whatever you specify. Takes either a function or
a configuration object.

- `middleware`: `function(req, res, next)`, is parsed like any other express middleware.
- `response.statusCode`: `number`, changes the statusCode of the response.
- `response.payload`: one of the following:
  - `string`, is sent in the body of the response.
  - `object`, is sent in the body of the response.
  
Examples:

```js
// with middleware
SecureRouter.denyWith(function (req, res, next) {
  // assuming AuthenticationError is defined elsewhere in your app
  // and handled later in another middleware that comes after bounceRequests
  return next(new AuthenticationError('Provide a site_token cookie!'));
}

// with a configuration object
SecureRouter.denyWith({
  statusCode: 404,
  payload: 'Nothing to see here, move along.'
});
```

### Defining bouncers on individual endpoints

Use the `secureEndpoint` method:

```js
router.secureEndpoint({
  method: 'GET',
  path: '/the-crown-jewels',
  bouncers: [bouncer1, bouncer2],
  middleware (req, res) {
    req.send(theCrownJewels);
  },
});
```

- **`method`**: `GET`, `POST`, `PUT`, `DELETE`, etc...
- **`path`**: appended to the path of the current router
- **`bouncers`**: list of bouncers. Alternatively, use **`bouncer`** to pass
  only one bouncer
- **`bouncer`**: individual bouncer. Alternatively, use **`bouncers`** to pass
  multiple bouncers
- **`middleware`**: either an array of middleware or a function. Works the
  same way as normal express middleware.

### Defining bouncers on subpaths

Use the `secureSubpath` method:

```js
let subpathRouter = router.secureSubpath({
  path: '/secrets',
  bouncers: [bouncer1, bouncer2],
});
```

Returns a new instance of `SecureRouter`.

- **`path`**: appended to the path of the current router
- **`bouncers`**: list of bouncers. Alternatively, use **`bouncer`** to pass
  only one bouncer
- **`bouncer`**: individual bouncer. Alternatively, use **`bouncers`** to pass
  multiple bouncers

### Defining flat bouncers

These bouncers apply to all routes used by the current router. Use the
`bouncer` method:

```
router.bouncer(bouncer);
```

## Contributing

```
npm install
npm run test
```


CHANGELOG.md
# [v3.0.0](https://github.com/goodeggs/secure-router/compare/v2.1.1...v3.0.0)

- If there is more than one bouncer that `DENY`s, the first one that is encountered
  is the one that will determine the response of the request.
  
- **BEWARE** - Bouncers are now expected to return a promise that resolves with an object
  that has a `middleware` property. Versions of `goodeggs-authentication-tokens` [<10.0.0](https://github.com/goodeggs/goodeggs-authentication-tokens/blob/master/CHANGELOG.md#v1000)
  use an earlier version of `sercure-router` in which `Router.denyWith` returns an object without the property.
  
  That means in addition to bumping secure-router you *must* bump any of the following module used in the project that use secure-router: 
  * `goodeggs-server>15`
  * `goodeggs-authentication-tokens>10`
  * `goodeggs-employee-roles>6`  

# [v2.0.0](https://github.com/goodeggs/secure-router/compare/v1.2.1...v2.0.0)

- `AUTHENTICATE`, `AUTHORIZE`, and `DENY` are now constants that live on the
  `SecureRouter` object.
- You can send custom responses when bouncing a request, with
  `SecureRouter.denyWith()`

```js
const bouncer = function (req, res) {
  getSecurityClearanceLevel(req.user.id)
  .then(function (level) {
    if (level === 'TOP SECRET') return SecureRouter.AUTHORIZE;
    else return SecureRouter.denyWith({
      statusCode: 404,
      payload: 'Nothing to see here, move along.'
    });
  });
}
```


test.js
/* eslint-env goodeggs/server-side-test */
import Promise from 'bluebird';
import _ from 'lodash';
import assert from 'assert';
import express from 'express';
import request from 'request';

import Router from './src/index';

describe('default behavior', function () {
  it('denies requests to endpoints with no security middleware', function () {
    const router = buildRouter();
    router.get('/foo', (req, res) => res.send('you should never get here!'));
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(401));
  });

  it('denies requests to undefined endpoints', function () {
    return withRunningServer(buildRouter())
    .then(() => expectRequest('GET', '/foo').toReturnCode(401));
  });
});

describe('use()', function () {
  it('allows nesting of secure routers with use()', function () {
    const router = buildRouter();
    const subRouter = new Router();
    subRouter.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    router.use(subRouter);
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(200));
  });
});

describe('secureEndpoint()', function () {
  it('allows GET access to specific secure endpoints', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(200))
    .then(() => expectRequest('POST', '/foo').toReturnCode(401))
    .then(() => expectRequest('GET', '/bar').toReturnCode(401));
  });

  it('allows POST access to specific secure endpoints', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'POST',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('POST', '/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/foo').toReturnCode(401))
    .then(() => expectRequest('POST', '/bar').toReturnCode(401));
  });

  it('allows PUT access to specific secure endpoints', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'PUT',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('PUT', '/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/foo').toReturnCode(401))
    .then(() => expectRequest('PUT', '/bar').toReturnCode(401));
  });

  it('allows DELETE access to specific secure endpoints', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'DELETE',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('DELETE', '/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/foo').toReturnCode(401))
    .then(() => expectRequest('DELETE', '/bar').toReturnCode(401));
  });

  it('allows HEAD access to specific secure endpoints', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'HEAD',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('HEAD', '/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/foo').toReturnCode(401))
    .then(() => expectRequest('HEAD', '/bar').toReturnCode(401));
  });

  it('allows requests with URL parameters', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo?bar=baz').toReturnCode(200));
  });

  it('ignores bouncers on other paths that do not share a secureSubpath', function () {
    const router = buildRouter();
    const fooRouter = new Router();

    router.use('/foo', fooRouter);
    fooRouter.secureEndpoint({
      method: 'POST',
      path: '/',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    fooRouter.secureEndpoint({
      method: 'POST',
      path: '/:id',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.denyWith({statusCode: 403})),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });

    return withRunningServer(router)
      .then(() => expectRequest('POST', '/foo').toReturnCode(200))
      .then(() => expectRequest('POST', '/foo/123').toReturnCode(403));
  });
});

describe('secureSubpath()', function () {
  it('allows access to sub-resources', function () {
    const router = buildRouter();
    router.secureSubpath({
      path: '/sub',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
    })
    .get('/foo', (req, res) => res.sendStatus(200));
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/sub/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/sub/bar').toReturnCode(404))
    .then(() => expectRequest('GET', '/bar').toReturnCode(401));
  });

  it('allows access to sub-sub-resources when the security is defined at the upper layer', function () {
    const router = buildRouter();
    const subSubRouter = new Router();
    subSubRouter.get('/foo', (req, res) => res.sendStatus(200));
    router.secureSubpath({
      path: '/sub',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
    })
    .use('/subsub', subSubRouter);
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/sub/subsub/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/sub/foo').toReturnCode(404));
  });

  it('allows access to sub-sub-resources when the security is defined at the lower layer', function () {
    const router = buildRouter();
    const subRouter = new Router();
    subRouter.secureSubpath({
      path: '/subsub',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
      ],
    })
    .get('/foo', (req, res) => res.sendStatus(200));
    router.use('/sub', subRouter);
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/sub/subsub/foo').toReturnCode(200))
    .then(() => expectRequest('GET', '/sub/foo').toReturnCode(401));
  });
});

describe('bouncer()', function () {
  it('allows you to arbitrarily add bouncers to routers', function () {
    const router = buildRouter();
    router.bouncer(_.constant(Router.AUTHENTICATE));
    router.bouncer(_.constant(Router.AUTHORIZE));
    router.get('/foo', (req, res) => res.sendStatus(200));
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(200));
  });
});

describe('allowing and denying', function () {
  it('does not allow access to a resource if no security middleware explicitly allows', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(),
        _.constant(),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(401));
  });

  it('does not allow access to a resource if some security middleware authenticates, some authorizes, and one denies', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
        _.constant(Router.DENY),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(401));
  });

  it('does allow access to a resource if some security middleware authenticates, some authorizes, and none denies', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(Router.AUTHORIZE),
        _.constant(),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(200));
  });

  it('does not allow access to a resource if the middleware does not authenticate (even if it authorizes)', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(),
        _.constant(Router.AUTHORIZE),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(401));
  });

  it('does not allow access to a resource (returns a 403) if the middleware authenticates but does not authorize', function () {
    const router = buildRouter();
    router.secureEndpoint({
      method: 'GET',
      path: '/foo',
      bouncers: [
        _.constant(Router.AUTHENTICATE),
        _.constant(),
      ],
      middleware: (req, res) => res.sendStatus(200),
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(403));
  });
});

describe('bouncer arguments', function () {
  it('provides req object as the first argument to the bouncer function', function () {
    const router = buildRouter();
    let url;
    router.bouncer(function (req) {
      url = req.url;
      return Router.AUTHENTICATE;
    });
    router.bouncer(_.constant(Router.AUTHORIZE));
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(404))
    .then(function () {
      assert(url.match(/\/foo/));
    });
  });

  it('provides res object as the second argument to the bouncer function', function () {
    const router = buildRouter();
    router.bouncer(function (req, res) {
      res.set('x-foo', 'bar');
      return Router.AUTHENTICATE;
    });
    router.bouncer(_.constant(Router.AUTHORIZE));
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toHaveHeader('x-foo', 'bar'));
  });
});

describe('custom denials with Router.denyWith()', function () {
  it('can run arbitrary middleware instead of responding', function () {
    const router = buildRouter();
    router.bouncer(function () {
      return Router.denyWith(function (req, res) {
        res.send(408);
      });
    });
    return withRunningServer(router)
    .then(function () {
      return expectRequest('GET', '/foo').toReturnCode(408);
    });
  });

  it('can provide custom status codes', function () {
    const router = buildRouter();
    router.bouncer(function () {
      return Router.denyWith({statusCode: 404});
    });
    return withRunningServer(router)
    .then(() => expectRequest('GET', '/foo').toReturnCode(404));
  });

  it('can provide a custom response payload as a string', function () {
    const router = buildRouter();
    router.bouncer(function () {
      return Router.denyWith({payload: 'Nope, sorry, not today!'});
    });
    return withRunningServer(router)
    .then(function () {
      return Promise.all([
        expectRequest('GET', '/foo').toReturnCode(401),
        expectRequest('GET', '/foo').toReturnBody('Nope, sorry, not today!'),
      ]);
    });
  });

  it('can provide a custom response payload as an object', function () {
    const router = buildRouter();
    router.bouncer(function () {
      return Router.denyWith({payload: {error: true, ms: 10239}});
    });
    return withRunningServer(router)
    .then(function () {
      return Promise.all([
        expectRequest('GET', '/foo').toReturnCode(401),
        expectRequest('GET', '/foo').toReturnBody({error: true, ms: 10239}),
      ]);
    });
  });

  it('applies denials in tree order (root first)', function () {
    /*
     * set up the following structure:
     *    / (bouncer1, bouncer2)
     *      /foo (bouncer3)
     *        /bar (bouncer4)
     */
    const router = buildRouter();
    const subRouter = new Router();
    const middlewareCalls = [];
    router.bouncer(function () {
      return Router.denyWith(function (req, res, next) {
        middlewareCalls.push(1);
        next();
      });
    });
    router.use('/foo', subRouter);
    subRouter.secureEndpoint({
      method: 'GET',
      path: '/bar',
      bouncer () {
        return Router.denyWith(function (req, res) {
          middlewareCalls.push(4);
          res.sendStatus(400);
        });
      },
      middleware: (req, res) => res.send(501),
    });
    router.bouncer(function () {
      return Router.denyWith(function (req, res, next) {
        middlewareCalls.push(2);
        next();
      });
    });
    subRouter.bouncer(function () {
      return Router.denyWith(function (req, res, next) {
        middlewareCalls.push(3);
        next();
      }, 3);
    });
    return withRunningServer(router)
    .then(function () {
      return expectRequest('GET', '/foo/bar').toReturnCode(400)
      .then(function () {
        assert.deepEqual(middlewareCalls, [1, 2, 3, 4]);
      });
    });
  });

  it('returns a 500 for an invalid bouncer (e.g. secure-router@2.1.1. Router.DENY)', function () {
    const router = buildRouter();

    router.bouncer(function () {
      // secure-router@2.1.1 returns this kind of bouncer if you use `Router.denyWith`
      return Promise.resolve({value: 'DENY', statusCode: 401});
    });

    router.secureEndpoint({
      method: 'GET',
      path: '/bar',
      middleware: (req, res) => res.send(200),
    });

    return withRunningServer(router)
    .then(function () {
      return expectRequest('GET', '/bar').toReturnCode(500);
    });
  });
});


function expectRequest (method, path) {
  const reqPromise = Promise.fromCallback(function (cb) {
    request({
      method,
      json: true,
      url: `http://localhost:19288${path}`,
    }, cb);
  });

  return {
    toReturnCode (responseCode) {
      return reqPromise
      .then(function (response) {
        assert.equal(response.statusCode, responseCode, `Expected request to ${path} to return code ${responseCode} but got ${response.statusCode}.`);
      });
    },
    toHaveHeader (name, value) {
      return reqPromise
      .then(function (response) {
        assert.equal(response.headers[name], value, `Expected request to ${path} to respond with header ${name} of value ${value}, but got ${response.headers[name]} instead.`);
      });
    },
    toReturnBody (body) {
      return reqPromise
      .then(function (response) {
        assert.deepEqual(response.body, body, `Expected request to ${path} to return body ${JSON.stringify(body)}, but got ${JSON.stringify(response.body)} instead.`);
      });
    },
  };
}

function buildRouter () {
  const router = new Router();
  router.bounceRequests();
  return router;
}

let sharedServer;
function withRunningServer (router) {
  const app = express();
  app.use(router);
  return Promise.fromCallback(function (done) {
    sharedServer = app.listen(19288, done);
  });
}
afterEach('stop server', function (done) {
  if (sharedServer) {
    return sharedServer.close(function () {
      sharedServer = null;
      done();
    });
  }
  return process.nextTick(done);
});


router.js
import BaseRouter from 'router';
import Promise from 'bluebird';
import _ from 'lodash';
import pathToRegexp from 'path-to-regexp';

export default class Router extends BaseRouter {
  constructor (...args) {
    super(...args);
    this.pathDefinitions = [];
    this.innerRouters = [];
  }

  getRequireOptInMiddleware () {
    const router = this;
    return function (req, res, next) {
      router.resolveCustomSecurity(req, req.url)
      .then(function (result) {
        if (result === 'ALLOW') return next();
        return res.sendStatus(405);
      });
    };
  }

  withSecurity (path, ...resolvers) {
    const innerRouter = new Router();
    this.pathDefinitions.push(createPathDefinition({path, innerRouters: [innerRouter], resolvers}));
    this.use(path, innerRouter);
    return innerRouter;
  }

  use (path, ...innerRouters) {
    if (_.isString(path)) {
      let {pathDefinition} = this.getPathDefinitionMatching(path);
      if (_.isNil(pathDefinition)) {
        pathDefinition = createPathDefinition({path});
        this.pathDefinitions.push(pathDefinition);
      }
      innerRouters = innerRouters.filter((router) => _.isFunction(router.resolveCustomSecurity));
      pathDefinition.innerRouters.push(...innerRouters);
    }
    BaseRouter.prototype.use.apply(this, arguments);
  }

  /* returns a promise. runs all of the appropriate middlewares configured for this route.
   *
   * if any return "DENY", then return the first "DENY".
   * if none return "DENY" and one returns "ALLOW", return "ALLOW".
   * otherwise return undefined.
   */
  resolveCustomSecurity (req, urlSegment) {
    return Promise.try(() => {
      const {pathDefinition, matchedUrlSegment} = this.getPathDefinitionMatching(urlSegment);
      if (!pathDefinition) return Promise.resolve();

      urlSegment = urlSegment.substr(matchedUrlSegment.length);

      return Promise.mapSeries(pathDefinition.resolvers, function (resolver) {
        return Promise.resolve(resolver(req))
        .then(function (result) {
          if (result === 'DENY') throw new Error();
          return result;
        });
      })
      .then(function (resultsForCurrentPath) {
        return Promise.mapSeries(pathDefinition.innerRouters, function (innerRouter) {
          return innerRouter.resolveCustomSecurity(req, urlSegment)
          .then(function (result) {
            if (result === 'DENY') throw new Error();
            return result;
          });
        })
        .then(function (resultsForInnerRouters) {
          return resultsForCurrentPath.concat(resultsForInnerRouters);
        });
      })
      .then(function (results) {
        if (_.some(results, (result) => result === 'ALLOW')) return 'ALLOW';
      })
      .catch(_.constant('DENY'));
    });
  }

  getPathDefinitionMatching (urlSegment) {
    let matches;
    for (const pathDefinition of this.pathDefinitions) {
      matches = pathDefinition.regexp.exec(urlSegment);
      if (_.isArray(matches) && matches.length > 0) {
        return {pathDefinition, matchedUrlSegment: matches[0]};
      }
    }
    return {pathDefinition: null};
  }
}

function createPathDefinition (definition) {
  return _.defaults({}, definition, {
    regexp: pathToRegexp(definition.path, {end: false, sensitive: true, strict: false}),
    innerRouters: [],
    resolvers: [],
  });
}


src/index.js
import BaseRouter from 'router';
import Promise from 'bluebird';
import _ from 'lodash';
import configureDebug from 'debug';
import http from 'http';
import pathToRegexp from 'path-to-regexp';
import url from 'url';
import async from 'async';

const debug = configureDebug('secure-router');

export default class Router extends BaseRouter {
  constructor (...args) {
    super(...args);
    this.bouncers = [];
    this.pathDefinitions = [];
  }

  static denyWith (params, i) {
    let middleware;

    if (_.isFunction(params)) {
      debug('Creating denial with a provided middleware function');
      middleware = params;
    } else if (_.isObject(params)) {
      debug('Creating denial with params', params);
      middleware = (req, res) => {
        debug('Responding to request with procedural denyWith middleware');
        res.status(_.get(params, 'statusCode', 401));
        res.send(params.payload);
      };
    }

    return _.assign({}, Router.DENY, {middleware, i});
  }

  bounceRequests () {
    const router = this;
    router.use(function (req, res, next) {
      const parsedUrl = url.parse(req.url);
      router.resolveCustomSecurity(req, res, parsedUrl.pathname)
      .then(function (results) {
        function match (result, expectedValue) {
          return _.isObject(result) && result.value === expectedValue;
        }

        function someResultsMatch (expectedValue) {
          return _.some(results, (result) => match(result, expectedValue));
        }
        function noResultsMatch (expectedValue) {
          return _.every(results, (result) => !match(result, expectedValue));
        }

        if (someResultsMatch(DENY)) {
          const denyResults = _.filter(results, (result) => match(result, DENY));
          debug('Got one or more DENY', {denyResults});
          // TODO(@max) sort here???
          async.eachSeries(
            _.map(denyResults, 'middleware'),
            (middleware, callback) => {
              try {
                middleware(req, res, callback);
              } catch (err) {
                debug('DENY middleware threw a synchronous error', err.stack);
                callback(err);
              }
            },
            next
          );
          return;
        }

        if (noResultsMatch(AUTHENTICATE)) {
          debug('Got no AUTHENTICATE from any bouncer, sending 401', {results});
          return res.sendStatus(401);
        }

        if (noResultsMatch(AUTHORIZE)) {
          debug('Got no AUTHORIZE from any bouncer, sending 403', {results});
          return res.sendStatus(403);
        }

        debug('Got one or more AUTHENTICATE and one or more AUTHORIZE', {results});
        return next();
      });
    });
    return router;
  }

  secureSubpath (params) {
    if (_.isNil(params.path)) throw new Error('Must pass a path to secureSubpath');

    let bouncers = params.bouncers;
    if (_.isNil(bouncers)) bouncers = [];
    if (_.isFunction(params.bouncer)) bouncers.push(params.bouncer);

    const innerRouter = new Router();
    this.pathDefinitions.push(createPathDefinition({
      path: params.path,
      innerRouters: [innerRouter],
      bouncers,
    }));
    this.use(params.path, innerRouter);
    return innerRouter;
  }

  secureEndpoint (params) {
    if (_.isNil(params.path)) throw new Error('Must pass a path to secureEndpoint');
    if (_.isNil(params.method)) throw new Error('Must pass a method to secureEndpoint');

    let bouncers = params.bouncers;
    if (_.isNil(bouncers)) bouncers = [];
    if (_.isFunction(params.bouncer)) bouncers.push(params.bouncer);

    let middleware = params.middleware;
    if (_.isNil(middleware)) throw new Error('Must pass a middlware to secureEndpoint');
    if (_.isFunction(middleware)) middleware = [middleware];

    this.pathDefinitions.push(createPathDefinition({
      path: params.path,
      bouncers,
      methods: [params.method.toUpperCase()],
    }));

    /* actually mount the route */
    this[params.method.toLowerCase()](params.path, ...middleware);
    return this;
  }

  use (path, ...innerRouters) {
    function isSecureRouter (router) {
      return _.isFunction(router.resolveCustomSecurity);
    }

    if (_.isString(path)) {
      let {pathDefinition} = this.getPathDefinitionMatching(path);
      if (_.isNil(pathDefinition)) {
        pathDefinition = createPathDefinition({path});
        this.pathDefinitions.push(pathDefinition);
      }
      innerRouters = innerRouters.filter(isSecureRouter);
      pathDefinition.innerRouters.push(...innerRouters);
    } else {
      innerRouters = [path, ...innerRouters].filter(isSecureRouter);
      for (const innerRouter of innerRouters) {
        this.bouncers.push(...innerRouter.bouncers);
        this.pathDefinitions.push(...innerRouter.pathDefinitions);
      }
    }
    BaseRouter.prototype.use.apply(this, arguments);
  }

  bouncer (bouncer) {
    this.bouncers.push(bouncer);
  }

  /* returns a promise. runs all of the appropriate bouncers configured for this route.  */
  resolveCustomSecurity (req, res, urlSegment) {
    return Promise.try(() => {
      const bouncers = [...this.bouncers];
      const innerRouters = [];

      const {pathDefinition, matchedUrlSegment} = this.getPathDefinitionMatching(urlSegment, req.method);
      debug('Found matching path definitions', {pathDefinition, matchedUrlSegment});
      if (_.isObject(pathDefinition)) {
        urlSegment = urlSegment.substr(matchedUrlSegment.length);
        bouncers.push(...pathDefinition.bouncers);
        innerRouters.push(...pathDefinition.innerRouters);
      }

      return Promise.all([
        Promise.map(bouncers, (bouncer) => Promise.resolve(bouncer(req, res))),
        Promise.map(innerRouters, (innerRouter) =>
          innerRouter.resolveCustomSecurity(req, res, urlSegment)
        ),
      ])
      .then(([bouncerResults, innerRouterResults]) => bouncerResults.concat(_.flatten(innerRouterResults)));
    });
  }

  getPathDefinitionMatching (urlSegment, method) {
    let matches, hasMethod;
    /* start with most specific path definition */
    this.pathDefinitions.sort((a, b) =>
      _.compact(b.path.split('/')).length - _.compact(a.path.split('/')).length
    );
    for (const pathDefinition of this.pathDefinitions) {
      matches = pathDefinition.regexp.exec(urlSegment);
      hasMethod = _.isNil(method) || _.includes(pathDefinition.methods, method);
      if (_.isArray(matches) && matches.length > 0 && hasMethod) {
        return {pathDefinition, matchedUrlSegment: matches[0]};
      }
    }
    return {pathDefinition: null, matchedUrlSegment: ''};
  }
}

const AUTHORIZE = 'AUTHORIZE';
const AUTHENTICATE = 'AUTHENTICATE';
const DENY = 'DENY';

Router.AUTHORIZE = {value: AUTHORIZE};
Router.AUTHENTICATE = {value: AUTHENTICATE};
Router.DENY = {
  value: DENY,
  middleware (req, res) {
    debug('Default DENY middleware responding with 401');
    res.sendStatus(401);
  },
};

function createPathDefinition (definition) {
  return _.defaults({}, definition, {
    regexp: pathToRegexp(definition.path, {end: false, sensitive: true, strict: false}),
    innerRouters: [],
    bouncers: [],
    methods: http.METHODS,
  });
}
